File: Top 100 Codes/01-Getting Started/18-Factorial of a number/Factorialofanumber.py
# Initialize the number for which we want to calculate the factorial
num = 6
fact = 1

# Check if the number is negative
if num < 0:
    # Factorial of a negative number is not defined
    print("Factorial of a negative number is not possible.")
else:
    # Calculate factorial using a loop
    for i in range(1, num + 1):
        fact *= i  # Multiply fact with the current number

    # Print the result
    print("Factorial of", num, "is", fact)

# Time Complexity: O(N) - Loop runs N times
# Space Complexity: O(1) - Constant space is used

File: Top 100 Codes/01-Getting Started/18-Factorial of a number/Factorialofanumber.cpp
#include <iostream>
using namespace std;

// Function to calculate the factorial of a given number recursively
int getFactorial(int num) {
    // Base case: factorial of 0 is 1
    if (num == 0)
        return 1;
    
    // Recursive call to calculate factorial
    return num * getFactorial(num - 1);
}

int main() {
    int num;

    // Prompt user for input
    cout << "Enter a number to find its factorial: ";
    cin >> num;

    // Validate input
    if (num < 0) {
        cout << "Factorial is not defined for negative numbers." << endl;
        return 1; // Exit with an error code
    }

    // Calculate factorial
    int fact = getFactorial(num);

    // Display result
    cout << "Factorial of " << num << " is: " << fact << endl;

    return 0;
}

// Time Complexity: O(N) - Linear recursion for N calls
// Space Complexity: O(1) - No extra space used apart from recursion stack
// Auxiliary Space Complexity (Function call stack): O(N)
